### PR TITLE
ide/docker: merge sandrinr/docker-xcsoar-buildenv into XCSoar

### PR DESCRIPTION
**Pull request by sandrinr**
Brief summary of the changes
----------------------------

- Use `delegated` volume modifier
- More concise notion of Docker commands
- Remove unnecessary "xcsoar" image path prefix
- Content and style improvements
- Support running XCSoar from within the Docker container

Related issues and discussions
------------------------------

This PR is an effort to semantically merge https://github.com/sandrinr/docker-xcsoar-buildenv into XCSoar.

### CODE REVIEW

File: .config/pulse/79372d8564c8545262975c46157f1482-runtime [hunk 1]
```diff
@@ -0,0 +1 @@
+/tmp/pulse-PKdhtXMmr18n
```
Comment (MaxKellermann) [inline .config/pulse/79372d8564c8545262975c46157f1482-runtime:1]: What's this?

File: .xcsoar/Default.tsk [hunk 2]
```diff
@@ -0,0 +1 @@
+<Task fai_finish="0" finish_min_height_ref="AGL" finish_min_height="0" start_max_height_ref="AGL" start_max_height="0" start_max_speed="0" start_requires_arm="0" aat_min_time="10800" type="RT"/>
```
Comment (MaxKellermann) [inline .xcsoar/Default.tsk:1]: What's this?

File: ide/docker/README.md [hunk 3]
```diff
@@ -47,3 +48,45 @@ docker run -it --rm -v $PWD:/opt/xcsoar:delegated xcsoar-build xcsoar-compile AN
...
 (macOS and Windows), it allows for eventual consistent filesystem change propagation 
 from the container to the host system.
+
+
+## Instructions for running XCSoar within the Docker container
+
+You can also use the Docker container to run the Linux build of XCSoar on a non-Linux
+system. This could be helpful to perform quick tests. It will use software rendering.
+Needless to say that the experience is inferior to running XCSoar on the host
+system directly. However, installing all build dependencies might also be overkill
```
Comment (MaxKellermann) [inline ide/docker/README.md:58]: You don't need to install build dependencies to run XCSoar. Only the shared libraries XCSoar was linked with. Which most users probably already have.

File: ide/docker/bin/xcsoar-compile [hunk 4]
```diff
@@ -19,13 +19,16 @@ case "${1}" in
...
+        make TARGET=UNIX OPENGL=n ENABLE_SDL=y USE_SDL2=y -j"${CPUS}" all
+      ;;
       "PC")
         make TARGET="${1}" -j"${CPUS}" all
       ;;
       "WIN64")
         make TARGET="${1}" -j"${CPUS}" all
       ;;
-      "*")
+      *)
```
Comment (MaxKellermann) [inline ide/docker/bin/xcsoar-compile:31]: How is this change related to running XCSoar inside docker?

File: .config/pulse/79372d8564c8545262975c46157f1482-runtime (hunk 1 above)
Comment (sandrinr) [reply .config/pulse/79372d8564c8545262975c46157f1482-runtime]: Commit error on my side. Removed.

File: .xcsoar/Default.tsk (hunk 2 above)
Comment (sandrinr) [reply .xcsoar/Default.tsk]: Commit error on my side. Removed.

File: ide/docker/README.md (hunk 3 above)
Comment (sandrinr) [reply ide/docker/README.md:58]: If you are running macOS or Windows you need a whole build environment, such as compilers.

File: ide/docker/bin/xcsoar-compile (hunk 4 above)
Comment (sandrinr) [reply ide/docker/bin/xcsoar-compile:31]: True, this is unrelated, however, it is still a bug in the current file. We can remove it here if you wish so.

Comment (rawtaz) [reply ide/docker/bin/xcsoar-compile:31]: +1 to removing it. Changes that are not related to this PR should not be in this PR. But feel free to open another PR for it if it makes sense to do so!

File: ide/docker/README.md (hunk 3 above)
Comment (MaxKellermann) [reply ide/docker/README.md:58]: Not for running.

File: ide/docker/bin/xcsoar-compile (hunk 4 above)
Comment (lordfolken) [reply ide/docker/bin/xcsoar-compile:31]: Please list the target UNIX-SDL also in the valid target lists on line 29

File: ide/docker/Dockerfile [hunk 5]
```diff
@@ -16,3 +16,10 @@ WORKDIR /opt/xcsoar
 
 COPY ./docker/bin/* /usr/local/bin/
 RUN cd /usr/local/bin && chmod 755 ./*
+
+# Environment to support running XCSoar
+ENV DISPLAY=host.docker.internal:0
```
Comment (lordfolken) [inline ide/docker/Dockerfile:21]: Hi, 
first of all thank you for following my call to action. :) +1 

Is this safe to hardcode for all possible runtime environments?

I saw for example this approach for a linux host: (the osx host will still require the above method)
https://medium.com/@SaravSun/running-gui-applications-inside-docker-containers-83d65c0db110

File: ide/docker/README.md (hunk 3 above)
Comment (sandrinr) [reply ide/docker/README.md:58]: Ah, now I got your point. Sure thing, but the target audience here is someone wanting to make (small) changes to XCSoar and therefore needing a way to compile it also.

I was in this situation some time ago and after spending some time failing to get it to work natively on macOS. I decided it would be easier to just let it run on the Docker container.

Comment (MaxKellermann) [reply ide/docker/README.md:58]: Sure they can compile it in the container, but the point of this PR is to *run* it in the container. You can compile in the container which has all the development tools installed, but is running inside the container really worth it?
It is dangerous and complex and error prone and slow; for example the user has to disable authentication in his X server.
(I'm not convinced, but that doesn't mean anything - I'm not convinced of Docker at all, that's why I have never used it.)

File: ide/docker/bin/xcsoar-compile (hunk 4 above)
Comment (sandrinr) [reply ide/docker/bin/xcsoar-compile]: done

File: ide/docker/README.md (hunk 3 above)
Comment (lordfolken) [reply ide/docker/README.md:58]: Is it possible to mount in the Xauthority file on OSX just like in the medium.com example? that would make it safer for the enduser.

@MaxKellermann docker is dead. there are much saner run-times for running containers now called cri-o and podman as the pulling tool. Its a challenge to run/compile master on pretty much anything that isn't debian buster or a direct target.

Comment (lordfolken) [reply ide/docker/README.md:58]: @sandrinr in theory one cloud build a runtime container only, or one only containing the unix target. It would be much smaller than this multigigabyte monster with all the toolchains. But for that kind of runtime I'd prefeer snapper or static binaries.

Comment (sandrinr) [reply ide/docker/README.md:58]: @MaxKellermann You got a point. However, AFAIK it is not possible to build macOS binaries on Linux. I think you need Xcode for that which is only available on macOS. So building the Linux binaries and running them within the Linux container is the whole point. Besides, in XCSoar history the Linux target always worked well. Not sth that could be told about the OSX target.
However, I do agree the approach being complicated (re-routing of X11 traffic, ...). It was just simpler for me as opposed to fixing the OSX target. I guess there is no harm in deciding that this should not be part of XCSoar.

File: ide/docker/Dockerfile (hunk 5 above)
Comment (sandrinr) [reply ide/docker/Dockerfile:21]: I already tried this. The article you mention is targeted towards Linux. I was unable to get this to run within reasonable time on macOS. The fact, that Docker runs within a virtual machine on macOS complicates things. The `DISPLAY` variable holds the path to a UNIX socket when using XQuartz.

Now that I am thinking about this again, the solution I provided here is only really suitable for macOS and also only because the OSX target is currently not working. (PC target is working I guess.)

Maybe it is not worth to put this all into XCSoar just to work around a defunct OSX target. Better to fix the OSX target or to use my buildenv as an alternative as long as it is defunct.

File: ide/docker/README.md (hunk 3 above)
Comment (rawtaz) [reply ide/docker/README.md:58]: @lordfolken Docker isn't dead, that's simply not true. It's used *a lot* all over the world. Even if there are better ones, they haven't hit mainstream use yet, and with that in mind having a docker container is better. It wouldn't hurt having others too, but let's not skip the docker one just because it's docker :)

Comment (MaxKellermann) [reply ide/docker/README.md:58]: So this is about building a Linux binary on Mac OS X and then running the Linux binary with Docker on Mac OS X? Does that even work? I thought Docker is a container manager, and containers share the host kernel, which isn't Linux... is Docker a VM on Mac OS X?

Comment (lordfolken) [reply ide/docker/README.md:58]: @MaxKellermann correct docker on os/x and windows use virtualization with very lightweight OSes (ramdisk/kernel)

Comment (rawtaz) [reply ide/docker/README.md:58]: Docker on macOS uses xhyve (practically bhyve) to run a VM one can enter into using `screen ~/Library/Containers/com.docker.docker/Data/vms/0/tty` to poke around:

~~~
docker-desktop:~# uname -a
Linux docker-desktop 4.19.76-linuxkit #1 SMP Thu Oct 17 19:31:58 UTC 2019 x86_64 Linux
docker-desktop:~# cat /etc/os-release 
PRETTY_NAME="Docker Desktop"
~~~

**Comment by lordfolken** [comment]
Please have a look at #428

**Comment by sandrinr** [comment]
We can close this since #428 was merged.
